titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/common/DifferenceResolverUtils.java

```diff
@@ -228,6 +233,51 @@ public static int countActiveNotStartedTasks(EntityHolder refJobHolder, EntityHo
         return pendingTaskIds.size();
     }
 
+    // Determines unassigned IP allocations based on the current reference view
+    public static Set<String> getUnassignedIpAllocations(JobView refJobView) {
+        // Get all IP allocations from the job
+        Set<String> unassignedIpAddressIds = refJobView.getJob().getJobDescriptor().getContainer().getContainerResources().getSignedIpAddressAllocations()
+                .stream()
+                .map(signedIpAddressAllocation -> signedIpAddressAllocation.getIpAddressAllocation().getAllocationId())
+                .collect(Collectors.toCollection(ConcurrentSkipListSet::new));
```

Comment:
After finishing the code review, I see that we can have many jobs with overlapping static IPs. There is a risk of deadlock for large jobs where not all tasks are created at once. For rolling push scenario, the old job may release IPs that are not assigned to tasks created in a new job.

Reply:
Yes, this is possible. The tasks that are created but not scheduled should end up with a task context indicating which running task they are waiting on to release the IP. This info should help a user triage the case you mentioned above.

AFAIK, all of the use cases for this feature are for relatively small jobs (e.g., a ZK cluster with 5 members). It may also make sense to put a relatively low limit on the max number of IP allocations that can be provided to a job (e.g., 10-20 or so), if we think it will get confusing for large jobs.

Reply:
Setting limit, and documenting it properly (how to avoid deadlock) should be ok.